### PR TITLE
fix: find experimental-ct-core cli as a fallback

### PR DESCRIPTION
**Pull request by mxschmitt**
https://github.com/microsoft/playwright/issues/29211

### CODE REVIEW

File: src/playwrightTest.ts [hunk 1]
```diff
@@ -63,11 +63,21 @@ export class PlaywrightTest {
...
+    } catch (error) {
+      return await this._getPlaywrightInfoImpl(workspaceFolder, configFilePath, '@playwright/experimental-ct-core');
+    }
+  }
+
+  private async _getPlaywrightInfoImpl(workspaceFolder: string, configFilePath: string, cliPackage: string): Promise<{ version: number, cli: string }> {
     const pwtInfo = await this._runNode([
       '-e',
-      'try { const pwtIndex = require.resolve("@playwright/test"); const version = require("@playwright/test/package.json").version; console.log(JSON.stringify({ pwtIndex, version})); } catch { console.log("undefined"); }',
+      `try { const pwtIndex = require.resolve("${cliPackage}"); const version = require("${cliPackage}/package.json").version; console.log(JSON.stringify({ pwtIndex, version})); } catch { console.log("null"); }`,
```
Comment (mxschmitt) [inline src/playwrightTest.ts:77]: Note: The reason for changing `undefined -> null` was that `JSON.parse('undefined')` does not work while it does with `null`. Added a null check afterwards to be on the safe side.

File: src/playwrightTest.ts [hunk 2]
```diff
@@ -64,10 +64,20 @@ export class PlaywrightTest {
   }
 
   async getPlaywrightInfo(workspaceFolder: string, configFilePath: string): Promise<{ version: number, cli: string }> {
+    try {
+      return await this._getPlaywrightInfoImpl(workspaceFolder, configFilePath, '@playwright/test');
+    } catch (error) {
+      return await this._getPlaywrightInfoImpl(workspaceFolder, configFilePath, '@playwright/experimental-ct-core');
```
Comment (mxschmitt) [inline src/playwrightTest.ts:70]:
```suggestion
      // In order to support component testing
      return await this._getPlaywrightInfoImpl(workspaceFolder, configFilePath, 'playwright');
```

File: src/playwrightTest.ts [hunk 3]
```diff
@@ -77,8 +87,10 @@ export class PlaywrightTest {
 
     const cliInfo = await this._runNode([
       '-e',
-      'try { const cli = require.resolve("@playwright/test/cli"); console.log(JSON.stringify({ cli })); } catch { console.log("undefined"); }',
+      `try { const path = require('path'); const cli = path.join(path.dirname(require.resolve("${cliPackage}")), 'cli.js'); console.log(JSON.stringify({ cli })); } catch { console.log("null"); }`,
```
[the remark this answers is not in the document]
Comment (dgozman) [reply src/playwrightTest.ts:91]: Perhaps this is not a problem with `playwright` package and we can revert to `require.resolve`?

**Review by dgozman — APPROVED** [review]

File: src/playwrightTest.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (mxschmitt) [reply src/playwrightTest.ts:91]: For future reference: cli.js is also not exported in the 'playwright' package.

**Comment by mxschmitt** [comment]
Tested in a React CT project and confirmed the fix.

**Pull request merged**